### PR TITLE
Native prices: Implement buffered batching coingecko 2/3 

### PR DESCRIPTION
**Pull request by m-lord-renkse**
# Description
Second part (2/3) on top of the PR: https://github.com/cowprotocol/services/pull/2869

Implements the trait `NativePriceBatchFetcher` for CoinGecko to allow bulk request.

# Changes
- Implements the trait `NativePriceBatchFetcher` for CoinGecko to allow bulk request
- Add new tests to ensure it is working as desired
- Refactor the CoinGecko implementation to adapt easily bulk request

## How to test
1. Unit test

### CODE REVIEW

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 1]
```diff
@@ -111,46 +128,93 @@ impl CoinGecko {
...
+impl NativePriceBatchFetcher for CoinGecko {
+    async fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> std::result::Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
+        let mut requested_tokens = tokens.iter().collect::<Vec<_>>();
+        match self.quote_token {
+            QuoteToken::Eth => self.send_bulk_request_price_in_eth(&requested_tokens).await,
+            QuoteToken::Other(native_price_token) => {
+                requested_tokens.push(&native_price_token);
```
Comment (squadgazzz) [inline crates/shared/src/price_estimation/native/coingecko.rs:164]: Any chance `native_price_token` already exists in the `requested_tokens`? No issue with sending requests with duplicated tokens?

Comment (Mateo-mro) [reply crates/shared/src/price_estimation/native/coingecko.rs:164]: It is not a problem to be twice, but actually there may be an issue with the limit. Fixing it. Thank you!

Comment (squadgazzz) [reply crates/shared/src/price_estimation/native/coingecko.rs:164]: So, does the batch size of the batched transport, in fact, control this limit? Say the coin gecko limit is 50. Should the batch size also take into account any additional tokens that might be added to the request?

Comment (Mateo-mro) [reply crates/shared/src/price_estimation/native/coingecko.rs:164]: Yup, that's exactly the issue I am fixing. Right now the limit is enforced by the PR 1/3 (https://github.com/cowprotocol/services/pull/2869), but this additional token may be a problem if the request already is already on its limit, so we should fire another request for the native price token.

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 2]
```diff
@@ -111,46 +128,93 @@ impl CoinGecko {
...
+#[async_trait]
+impl NativePriceBatchFetcher for CoinGecko {
+    async fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> std::result::Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
+        let mut requested_tokens = tokens.iter().collect::<Vec<_>>();
+        match self.quote_token {
+            QuoteToken::Eth => self.send_bulk_request_price_in_eth(&requested_tokens).await,
+            QuoteToken::Other(native_price_token) => {
```
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/coingecko.rs:163]: nit: the `Other` code path is quite big. Might be nicer to have it in a separate function to improve readability.

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 3]
```diff
@@ -111,46 +128,93 @@ impl CoinGecko {
...
+                            let token_in_native_price_f64 = token_in_native_price.to_f64().ok_or(
+                                PriceEstimationError::EstimatorInternal(anyhow!(
+                                    "failed to parse result to f64"
+                                )),
+                            )?;
+                            Ok((token, Ok(token_in_native_price_f64)))
+                        }
+                        Err(_) => Ok((token, price)),
+                    })
+                    .collect::<Result<_, _>>()
```
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/coingecko.rs:204]: Collecting into a single `Result` means a single result will invalidate the whole batch. Only omitting the bad prices and returning all the good ones seems like the better choice.

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 4]
```diff
@@ -111,46 +128,93 @@ impl CoinGecko {
...
 
-        let price = prices
+        Ok(tokens
+            .iter()
+            .map(|token| {
+                (
+                    **token,
+                    prices
+                        .get(token)
+                        .ok_or(PriceEstimationError::NoLiquidity)
```
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/coingecko.rs:138]: Not sure the `NoLiquidity` case needs to be handled at this low level. There are multiple code paths that convert a missing token price lookup into such an error so I think you could basically just return the coingecko response here.

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 5]
```diff
@@ -90,7 +102,12 @@ impl CoinGecko {
         if !response.status().is_success() {
             let status = response.status();
             return match status {
-                StatusCode::TOO_MANY_REQUESTS => Err(PriceEstimationError::RateLimited),
+                StatusCode::TOO_MANY_REQUESTS => {
+                    return Ok(tokens
+                        .iter()
+                        .map(|token| (**token, Err(PriceEstimationError::RateLimited)))
```
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/coingecko.rs:108]: Why do you return a rate limiting error for every token here instead of one single `Err` from the function?
That seems to describe better what kind of error actually happened. 🤔

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 6]
```diff
@@ -71,12 +72,23 @@ impl CoinGecko {
         })
     }
 
-    pub async fn send_request_price_in_eth(&self, token: Token) -> NativePriceEstimateResult {
+    pub async fn send_bulk_request_price_in_eth(
```
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/coingecko.rs:75]: There are a few function that are public but seem to be only implementation details. Can we make them private instead?

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 4 above)
Comment (Mateo-mro) [reply crates/shared/src/price_estimation/native/coingecko.rs:138]: Here we are converting the raw CoinGecko response into a domain  type `(H16,  NativePriceResult)`. What we do here is check if we got an answer for each token requested, and for those without an answer, we return `NoLiquidity`. CoinGecko works that way, if there is no answer for the token requested, it is because there is no liquidty.

We cannot pass the CoinGecko response directly, we need to standarize it to our internal domain type first.

Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:138]: The point I'm trying to make is that you have to handle that `map.get(token)` returns `None` in any case. So I think it's should be perfectly fine to just have the convention that not finding the token is equivalent to finding it with the error `NoLiquidity`.
That way you could simply return the `Response` here without having to do any transformations.

File: crates/shared/src/price_estimation/buffered.rs [hunk 7]
```diff
@@ -35,41 +33,36 @@ pub struct Configuration {
     pub max_batch_len: usize,
     /// An additional minimum delay to wait for collecting requests.
     ///
-    /// The delay to start counting after receiving the first request.
-    pub debouncing_time: Duration,
+    /// The delay starts counting after receiving the first request.
+    pub batch_delay: Duration,
```
Comment (fleupold) [inline crates/shared/src/price_estimation/buffered.rs:37]: Are these changes intended or missing a rebase?

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 8]
```diff
@@ -71,101 +72,167 @@ impl CoinGecko {
...
+        Ok(parsed_response)
+    }
+
+    /// Fetches the prices of the given tokens denominated in the given token.
+    async fn bulk_fetch_denominated_in_token(
+        &self,
+        mut tokens: Vec<&Token>,
+        denominator: Token,
+    ) -> Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
+        tokens.push(&denominator);
```
Comment (fleupold) [inline crates/shared/src/price_estimation/native/coingecko.rs:129]: How do we make sure we don't make the list of tokens too large? Careful configuration of the batcher (ie. for xDAI max batch size needs to be 19 instead of 20?).

Alternatively, the `NativePriceBatchFetcher` trait could expose a method `batchSize` which would allow to colocate the logic of how big a batch can be with the implementation of the request (and here we could depending on the weth token symbol return 19 or 20 batch size.

**Review by fleupold — COMMENTED** [review]
This PR likely needs to be rebased to incorporate all the improvements of 1/3

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 8 above)
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:129]: > Alternatively, the NativePriceBatchFetcher trait could expose a method batchSize

I played around with exactly that and it didn't seem too great given that we already have this as a config value.

File: crates/shared/src/price_estimation/buffered.rs (hunk 7 above)
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/buffered.rs:37]: Not intended. Must have messed up when rebasing. :/

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 8 above)
Comment (Mateo-mro) [reply crates/shared/src/price_estimation/native/coingecko.rs:129]: If `tokens` is 20, then this query would fail, because appending extra token will make CoinGecko query invalid (I don't know how it behaves in that case, if it ignores the extra token or directly fails).

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 9]
```diff
@@ -71,101 +72,167 @@ impl CoinGecko {
         })
     }
 
-    pub async fn send_request_price_in_eth(&self, token: Token) -> NativePriceEstimateResult {
+    async fn bulk_fetch_denominated_in_eth(
+        &self,
+        tokens: &[&Token],
+    ) -> Result<HashMap<Token, f64>, PriceEstimationError> {
         let mut url = crate::url::join(&self.base_url, &self.chain);
```
Comment (Mateo-mro) [inline crates/shared/src/price_estimation/native/coingecko.rs:79]: We can have a size check here for `tokens` with an early return is exceed the bulk size.

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 10]
```diff
@@ -71,101 +72,167 @@ impl CoinGecko {
...
+                            .get(token)
+                            .cloned()
+                            .ok_or(PriceEstimationError::NoLiquidity);
+                        (*token, result)
+                    })
+                    .collect())
+            }
+            QuoteToken::Other(native_price_token) => {
+                if !requested_tokens.contains(&native_price_token) {
+                    tokens.push(&native_price_token);
```
Comment (Mateo-mro) [inline crates/shared/src/price_estimation/native/coingecko.rs:197]: careful here with the `token` size

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 11]
```diff
@@ -176,3 +243,102 @@ mod observe {
...
+            .fetch_native_prices(&HashSet::from([usdt_token, usdc_token]))
+            .await
+            .unwrap();
+        let usdt_price = estimated_price.get(&usdt_token).unwrap().clone();
+        let usdc_price = estimated_price.get(&usdc_token).unwrap().clone();
+        // Since the USDT/USDC precise price against XDAI is not always exact to
+        // 1.0 (it can vary slightly)
+        assert!((0.95..=1.05).contains(&usdt_price.unwrap()));
+        assert!((0.95..=1.05).contains(&usdc_price.unwrap()));
+    }
```
Comment (Mateo-mro) [inline crates/shared/src/price_estimation/native/coingecko.rs:343]: worth adding a test to check the bulk limit too

**Comment by Mateo-mro** [comment]
@MartinquaXD can you change this code:

```rust
struct Price {
    eth: f64,
}
```

to 

```rust
struct Price {
    eth: Option<f64>,
}
```

It is a mistake on my side (since I wrote that). The reason is that when CoinGecko doesn't know liquidity, it returns the hashmap but with no `eth`. So while on-call I saw we error here, not a big deal since a parsing error ends up in `NoLiquidity` too, but it is not the most elegant way.

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 9 above)
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:79]: Initially thought about that but ultimately I think this is a configuration error and you will get an error anyway when sending the big request.

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 10 above)
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:197]: Yes, I'd say the responsibility lies on the config to provide 19 as the batch size.

**Comment by Mateo-mro** [comment]
Sorry, I think it should be:

```rust
struct Response(HashMap<H160, Option<Price>>);
```

Rather than my change proposed above, please, check in the test :pray: thank you!!

File: crates/shared/src/price_estimation/buffered.rs [hunk 12]
```diff
@@ -52,13 +50,18 @@ pub trait NativePriceBatchFetching: Sync + Send + NativePriceEstimating {
     ///
     /// It returns a HashMap which maps the token with its native price
     /// estimator result
-    fn fetch_native_prices(
-        &self,
-        tokens: &HashSet<H160>,
+    fn fetch_native_prices<'a, 'b>(
+        &'a self,
+        tokens: &'b HashSet<H160>,
```
Comment (fleupold) [inline crates/shared/src/price_estimation/buffered.rs:55]: To avoid annoying life times maybe we should pass this `HashSet` by value (probably isn't a performance isse in practice)?

**Review by fleupold — APPROVED** [review]

**Comment by MartinquaXD** [comment]
> It is a mistake on my side (since I wrote that). The reason is that when CoinGecko doesn't know liquidity, it returns the hashmap but with no eth. So while on-call I saw we error here, not a big deal since a parsing error ends up in NoLiquidity too, but it is not the most elegant way.

This is not a big deal for single token requests but that parsing error would also cause a whole batch of good prices to be discarded. Added the fix and a unit test to verify that we still get all the known prices even with a nonsense token in the batch.

File: crates/shared/src/price_estimation/buffered.rs [hunk 13]
```diff
@@ -158,38 +159,43 @@ where
...
-                        .into_iter()
-                        .map(|(token, price)| NativePriceResult {
-                            token,
-                            result: price,
-                        })
-                        .collect(),
-                    Err(err) => {
-                        tracing::error!(?err, "failed to send native price batch request");
-                        batch
+        tokio::task::spawn(batched_for_each(
```
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/buffered.rs:162]: Only reformatted due to the new argument.

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 11 above)
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:361]: Not sure about that since this is nothing we can prevent given the current setup anyway.
To reasonably enforce that in a way the can be tested the `BufferedPriceFetching` would have to spawn its own background task with the enforced limit.
At the moment this gets set up outside of this component so even if we have a test that detects some sort of error if `batch_size > 20` it would not help us prevent this error in the actual business logic.

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 14]
```diff
@@ -71,101 +75,182 @@ impl CoinGecko {
...
+        PriceEstimationError::EstimatorInternal(err)
+    }
 }
 
 mod observe {
     use url::Url;
 
     /// Observe a request to be sent to CoinGecko
-    pub fn coingecko_request(endpoint: &Url) {
+    pub(super) fn coingecko_request(endpoint: &Url) {
```
Comment (m-lord-renkse) [inline crates/shared/src/price_estimation/native/coingecko.rs:248]: btw, fyi, this is redundant (no need to change, out of curiosity), `observe` is only accessible within the `coingecko.rs` file anyway.

Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:248]: Yeah, I'm just trying to be conscious about the visibility of items to help the compiler find dead code and avoid leaking things into the public interface of a crate. Once things are there the compiler can't help us find dead code anymore. :/

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 15]
```diff
@@ -71,101 +75,182 @@ impl CoinGecko {
...
-                        token_eth.checked_div(native_price_token_eth).ok_or(
-                            PriceEstimationError::EstimatorInternal(anyhow!("division by zero")),
-                        )?;
-                    token_in_native_price
-                        .to_f64()
-                        .ok_or(PriceEstimationError::EstimatorInternal(anyhow!(
-                            "failed to parse result to f64"
-                        )))
+                    if !requested_tokens.contains(&native_price_token) {
+                        tokens.push(&native_price_token);
```
Comment (sunce86) [inline crates/shared/src/price_estimation/native/coingecko.rs:200]: Are we pushing denominator here and then again inside `bulk_fetch_denominated_in_token`?

Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs]: We do. Good catch!
Fixed now.

**Review by sunce86 — APPROVED** [review]

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 14 above)
Comment (m-lord-renkse) [reply crates/shared/src/price_estimation/native/coingecko.rs:243]: That's true, you are totally right

**Review by squadgazzz — APPROVED** [review]
LGTM
